### PR TITLE
Rewrite user stylesheet injection for WebEngine

### LINKED ISSUE #2723: userscript:_qute_stylesheet: Uncaught TypeError: Cannot read property 'appendChild' of undefined
I am seeing the following error when accessing https://travis-ci.com/private_org/private_repo/pull_requests:

> 14:07:36 ERROR: [userscript:_qute_stylesheet:6] Uncaught TypeError: Cannot read property 'appendChild' of undefined

Before there was a warning already:

> 14:07:32 WARNING: [https://cdn.travis-ci.com/assets/vendor-284b882d99e4c1b0d34e3a66b48188d1.js:21] Error

```
% qutebrowser --version
qutebrowser v0.10.1
Git commit: 
Backend: QtWebEngine

CPython: 3.6.1
Qt: 5.9.0
PyQt: 5.8.2

sip: 4.19.2
colorama: 0.3.9
pypeg2: 2.15
jinja2: 2.9.6
pygments: 2.2.0
yaml: 3.12
cssutils: no
typing: yes
PyQt5.QtWebEngineWidgets: yes
pdf.js: 1.6.210 (/usr/share/pdf.js/build/pdf.js)
Webkit: 538.1
SSL: OpenSSL 1.0.2l  25 May 2017

Style: QFusionStyle
Platform: Linux-4.11.5-1-ARCH-x86_64-with-arch, 64bit
Frozen: False
Imported from /usr/lib/python3.6/site-packages/qutebrowser
Qt library executable path: /usr/lib/qt/libexec, data path: /usr/share/qt
OS Version: 

--- /etc/lsb-release ---
LSB_VERSION=1.4
DISTRIB_ID=Arch
DISTRIB_RELEASE=rolling
DISTRIB_DESCRIPTION="Arch Linux"

--- /etc/os-release ---
NAME="Arch Linux"
PRETTY_NAME="Arch Linux"
ID=arch
ID_LIKE=archlinux

Paths:
config: /home/daniel/.local/share/qutebrowser-work/config
data: /home/daniel/.local/share/qutebrowser-work/data
system_data: /usr/share/qutebrowser
cache: /home/daniel/.local/share/qutebrowser-work/cache
download: /home/daniel/.local/share/qutebrowser-work/download
runtime: /home/daniel/.local/share/qutebrowser-work/runtime
```

I assume it comes from https://github.com/qutebrowser/qutebrowser/blob/2ffb10badf723522d82256154716c8cd344dec4a/qutebrowser/browser/webengine/webenginesettings.py#L145-L152 somehow?!

(the `cache` dir does not exist currently (it is a symlink pointing to a non-existing dir), in case that matters)

### LINKED ISSUE #2771: Make user-stylesheet work better with QtWebEngine
After [QTBUG-51192](https://bugreports.qt.io/browse/QTBUG-51192) was abandoned, I guess it's time to see if we can fix user stylesheets in some different way.

Looking at [Stylus](https://github.com/openstyles/stylus) (Stylish-fork which is open source), it seems to do some pretty [complex stuff](https://github.com/openstyles/stylus/blob/master/apply.js) to avoid flickering, but it doesn't actually need much of the Chromium WebExtension API, so it might be possible to do the same in qutebrowser.

### PR DESCRIPTION
**Pull request by ulrikdem**
This now works correctly in XML documents. The stylesheet is applied at document creation to reduce flickering, and is updated if the `user_stylesheets` setting is changed after page load. Fixes #2723 and closes #2771.

See #2935 for the discussion of my previous attempt.

This works based on manual testing. Automated tests have not yet been written.

---
This change is [Reviewable](https://reviewable.io/reviews/qutebrowser/qutebrowser/3212)

### CODE REVIEW

**Comment by The-Compiler** [comment]
Thanks!

@sandrosc Can you review the [JavaScript part](https://github.com/qutebrowser/qutebrowser/pull/3212/files#diff-245d3bf95eaf3a8a7d4c26bc14710e9a)?
@jgkamat Do you still want to write tests for this? :wink:

**Comment by sandrosc** [comment]
Review status: 0 of 4 files reviewed at latest revision, 5 unresolved discussions.

---

*[qutebrowser/javascript/stylesheet.js, line 48 at r1](https://reviewable.io:443/reviews/qutebrowser/qutebrowser/3212#-KxgPIjjDHQP2BSpbuV7:-KxgPIjjDHQP2BSpbuV8:b-dr8ie7) ([raw file](https://github.com/qutebrowser/qutebrowser/blob/2fe1a1db89f97c23b3ca1aeac8fb33f4d8c5f940/qutebrowser/javascript/stylesheet.js#L48)):*
> ```JavaScript
>             root_observer.disconnect();
>             root_observer.observe(document, {"childList": true});
>             root_observer.observe(root_elem, {"childList": true});
> ```

As far as I can tell we don't need to watch `document.documentElement`, since we're watching `document` already.
Any particular reason for this?

---

*[qutebrowser/javascript/stylesheet.js, line 56 at r1](https://reviewable.io:443/reviews/qutebrowser/qutebrowser/3212#-KxgHQkejF0UBOmNMdqM:-KxgHQkejF0UBOmNMdqN:bs67fcu) ([raw file](https://github.com/qutebrowser/qutebrowser/blob/2fe1a1db89f97c23b3ca1aeac8fb33f4d8c5f940/qutebrowser/javascript/stylesheet.js#L56)):*
> ```JavaScript
> 
>     function create_style() {
>         var ns = xhtml_ns;
> ```

This can also be a conditional assignment. `var ns = nsURI === svg_ns ? svg_ns : xhtml_ns`.

---

*[qutebrowser/javascript/stylesheet.js, line 71 at r1](https://reviewable.io:443/reviews/qutebrowser/qutebrowser/3212#-KxgMw8XR2RutzWEHUL5:-KxgMw8XR2RutzWEHUL6:b-7qckpm) ([raw file](https://github.com/qutebrowser/qutebrowser/blob/2fe1a1db89f97c23b3ca1aeac8fb33f4d8c5f940/qutebrowser/javascript/stylesheet.js#L71)):*
> ```JavaScript
>     // starting point for exploring the relevant code in Chromium, see
>     // https://github.com/qt/qtwebengine-chromium/blob/cfe8c60/chromium/third_party/WebKit/Source/core/xml/parser/XMLDocumentParser.cpp#L1539-L1540
>     function check_style(node) {
> ```

This method calls `create_style` if certain preconditions are met. The `check_style` name is a bit misleading.

---

*[qutebrowser/javascript/stylesheet.js, line 85 at r1](https://reviewable.io:443/reviews/qutebrowser/qutebrowser/3212#-KxgQjeufUL_9xB6IWF6:-KxgLLyNpysH5N0Kl_IJ:bpvkvbz) ([raw file](https://github.com/qutebrowser/qutebrowser/blob/2fe1a1db89f97c23b3ca1aeac8fb33f4d8c5f940/qutebrowser/javascript/stylesheet.js#L85)):*
> ```JavaScript
>     }
> 
>     function check_added_style(mutations) {
> ```

This function is called `check_...`, but returns `undefined`.

---

*[qutebrowser/javascript/stylesheet.js, line 105 at r1](https://reviewable.io:443/reviews/qutebrowser/qutebrowser/3212#-KxgGWIbkz_CflTlPlMV:-KxgGWIbkz_CflTlPlMW:bhawrxg) ([raw file](https://github.com/qutebrowser/qutebrowser/blob/2fe1a1db89f97c23b3ca1aeac8fb33f4d8c5f940/qutebrowser/javascript/stylesheet.js#L105)):*
> ```JavaScript
>             return;
>         }
>         var iter = document.createNodeIterator(document);
> ```

For which elements should `check_style` be called? `createNodeIterator` allows for some filtering, but could we also define a list of elements to be checked?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/qutebrowser/qutebrowser/3212)*

**Comment by jgkamat** [comment]
@The-Compiler, I'm actually a bit busy this week (I have a huge project deadline coming up), but I would be happy to work on it sometime next week. 

I'm not sure exactly how I would even approach these tests though, so I'll probably be guessing and checking how to start off :P

**Comment by The-Compiler** [comment]
I hope everything goes well with that project! If you have some qutebrowser time left in between things, I'd say #3181 and #3122 are more important FWIW 😄 

`tests/unit/javascript/position_caret/test_position_caret.py` might be a good starting point, as it implements unit tests for JS code. 

It uses `JSTester` from `tests/unit/javascript/conftest.py` as a `js_tester` fixture, which is only used there so far, but is general enough to hopefully help with testing other JS files like this as well.

**Comment by The-Compiler** [comment]
Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[qutebrowser/browser/webengine/webenginesettings.py, line 167 at r1](https://reviewable.io:443/reviews/qutebrowser/qutebrowser/3212#-KxiAL-Gecit57wx8Pj6:-KxiAL-Gecit57wx8Pj7:b-km0xse) ([raw file](https://github.com/qutebrowser/qutebrowser/blob/2fe1a1db89f97c23b3ca1aeac8fb33f4d8c5f940/qutebrowser/browser/webengine/webenginesettings.py#L167)):*
> ```Python
>         '"use strict";',
>         'window._qutebrowser = window._qutebrowser || {};',
>         utils.read_file('javascript/stylesheet.js'),
> ```

I guess there isn't a straightforward way to only inject this file once (but early enough), like other files are initialized? Is `WebEngineTab._init_js` in `qutebrowser/browser/webengine/webenginetab.py` too late?

---

*[qutebrowser/browser/webengine/webenginesettings.py, line 184 at r1](https://reviewable.io:443/reviews/qutebrowser/qutebrowser/3212#-KxiAlEsjMsQ_jBtRaBf:-KxiAlEsjMsQ_jBtRaBg:b-a0npk2) ([raw file](https://github.com/qutebrowser/qutebrowser/blob/2fe1a1db89f97c23b3ca1aeac8fb33f4d8c5f940/qutebrowser/browser/webengine/webenginesettings.py#L184)):*
> ```Python
>     css = shared.get_user_stylesheet()
>     code = javascript.assemble('stylesheet', 'set_css', css)
>     for win_id in objreg.window_registry:
> ```

This will probably fail when a window is in the process of shutting down - this should help:

```python
        # We could be in the middle of destroying a window here
        if sip.isdeleted(window):
            continue
```

---

*[qutebrowser/javascript/stylesheet.js, line 56 at r1](https://reviewable.io:443/reviews/qutebrowser/qutebrowser/3212#-KxgHQkejF0UBOmNMdqM:-Kxi9isc2jgYwkkBUAxV:b-eej3am) ([raw file](https://github.com/qutebrowser/qutebrowser/blob/2fe1a1db89f97c23b3ca1aeac8fb33f4d8c5f940/qutebrowser/javascript/stylesheet.js#L56)):*
<details><summary><i>Previously, sandrosc (Sandro) wrote…</i></summary><blockquote>

This can also be a conditional assignment. `var ns = nsURI === svg_ns ? svg_ns : xhtml_ns`.
</blockquote></details>

Though you'd need to define `nsURI` just for that check (or have a rather long one-liner)
var ns = nsURI === svg_ns ? svg_ns : xhtml_ns

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/qutebrowser/qutebrowser/3212)*

**Comment by The-Compiler** [comment]
*[qutebrowser/javascript/stylesheet.js, line 56 at r1](https://reviewable.io:443/reviews/qutebrowser/qutebrowser/3212#-KxgHQkejF0UBOmNMdqM:-KxiBiHhbic7pTsx3ZD6:bx3gde3) ([raw file](https://github.com/qutebrowser/qutebrowser/blob/2fe1a1db89f97c23b3ca1aeac8fb33f4d8c5f940/qutebrowser/javascript/stylesheet.js#L56)):*
<details><summary><i>Previously, The-Compiler (Florian Bruhin) wrote…</i></summary><blockquote>

Though you'd need to define `nsURI` just for that check (or have a rather long one-liner)
var ns = nsURI === svg_ns ? svg_ns : xhtml_ns
</blockquote></details>

Whoops, I messed up that comment - I wanted to say "so it's probably okay as-is" :wink:

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/qutebrowser/qutebrowser/3212)*

**Comment by ulrikdem** [comment]
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[qutebrowser/browser/webengine/webenginesettings.py, line 167 at r1](https://reviewable.io:443/reviews/qutebrowser/qutebrowser/3212#-KxiAL-Gecit57wx8Pj6:-KxiG1H-7UoTXJVRPHSD:b-vpkjyt) ([raw file](https://github.com/qutebrowser/qutebrowser/blob/2fe1a1db89f97c23b3ca1aeac8fb33f4d8c5f940/qutebrowser/browser/webengine/webenginesettings.py#L167)):*
<details><summary><i>Previously, The-Compiler (Florian Bruhin) wrote…</i></summary><blockquote>

I guess there isn't a straightforward way to only inject this file once (but early enough), like other files are initialized? Is `WebEngineTab._init_js` in `qutebrowser/browser/webengine/webenginetab.py` too late?
</blockquote></details>

It is too late. However, wouldn't the JS file be read once per tab in `WebEngineTab._init_js` but only when the configuration is changed where it is currently? And injected once into each page in either case?

---

*[qutebrowser/browser/webengine/webenginesettings.py, line 184 at r1](https://reviewable.io:443/reviews/qutebrowser/qutebrowser/3212#-KxiAlEsjMsQ_jBtRaBf:-KxiJ2H8odsnu3x3PgQm:b-xwfeeh) ([raw file](https://github.com/qutebrowser/qutebrowser/blob/2fe1a1db89f97c23b3ca1aeac8fb33f4d8c5f940/qutebrowser/browser/webengine/webenginesettings.py#L184)):*
<details><summary><i>Previously, The-Compiler (Florian Bruhin) wrote…</i></summary><blockquote>

This will probably fail when a window is in the process of shutting down - this should help:

```python
        # We could be in the middle of destroying a window here
        if sip.isdeleted(window):
            continue
```
</blockquote></details>

I will add this check.

---

*[qutebrowser/javascript/stylesheet.js, line 48 at r1](https://reviewable.io:443/reviews/qutebrowser/qutebrowser/3212#-KxgPIjjDHQP2BSpbuV7:-Kxhrtt9BNIDtm9JdHen:bsz36cu) ([raw file](https://github.com/qutebrowser/qutebrowser/blob/2fe1a1db89f97c23b3ca1aeac8fb33f4d8c5f940/qutebrowser/javascript/stylesheet.js#L48)):*
<details><summary><i>Previously, sandrosc (Sandro) wrote…</i></summary><blockquote>

As far as I can tell we don't need to watch `document.documentElement`, since we're watching `document` already.
Any particular reason for this?
</blockquote></details>

`observe` does not recursively observe the node's children, unless the `subtree` option is given. `document` is observed to detect when `documentElement` is replaced, and `documentElement` is observed to detect when its children are added or removed. The latter allows the `style` element to be re-added if it is removed, and is necessary to move it to the end of the document. This is an idea I borrowed from Stylus. It is not strictly necessary, but the element feels more out of the way there, and its rules take precedence over other stylesheets (among selectors with the same specificity).

---

*[qutebrowser/javascript/stylesheet.js, line 56 at r1](https://reviewable.io:443/reviews/qutebrowser/qutebrowser/3212#-KxgHQkejF0UBOmNMdqM:-Kxhtn70UfU2aLUaRWMQ:bi3ly72) ([raw file](https://github.com/qutebrowser/qutebrowser/blob/2fe1a1db89f97c23b3ca1aeac8fb33f4d8c5f940/qutebrowser/javascript/stylesheet.js#L56)):*
<details><summary><i>Previously, The-Compiler (Florian Bruhin) wrote…</i></summary><blockquote>

Whoops, I messed up that comment - I wanted to say "so it's probably okay as-is" :wink:
</blockquote></details>

I had that originally, but ESLint doesn't like the ternary operator with its current configuration.

---

*[qutebrowser/javascript/stylesheet.js, line 71 at r1](https://reviewable.io:443/reviews/qutebrowser/qutebrowser/3212#-KxgMw8XR2RutzWEHUL5:-Kxi7o6SjFqbkDW7G9WQ:b-5aonjp) ([raw file](https://github.com/qutebrowser/qutebrowser/blob/2fe1a1db89f97c23b3ca1aeac8fb33f4d8c5f940/qutebrowser/javascript/stylesheet.js#L71)):*
<details><summary><i>Previously, sandrosc (Sandro) wrote…</i></summary><blockquote>

This method calls `create_style` if certain preconditions are met. The `check_style` name is a bit misleading.
</blockquote></details>

I will move the `create_style` call out of this function.

---

*[qutebrowser/javascript/stylesheet.js, line 85 at r1](https://reviewable.io:443/reviews/qutebrowser/qutebrowser/3212#-KxgQjeufUL_9xB6IWF6:-Kxi8_ZTzjdy205ATD2j:b-5jwlzl) ([raw file](https://github.com/qutebrowser/qutebrowser/blob/2fe1a1db89f97c23b3ca1aeac8fb33f4d8c5f940/qutebrowser/javascript/stylesheet.js#L85)):*
<details><summary><i>Previously, sandrosc (Sandro) wrote…</i></summary><blockquote>

This function is called `check_...`, but returns `undefined`.
</blockquote></details>

How about `watch_added_style`? I would have implemented this as a function expression passed to the `MutationObserver` constructor. ESLint suggests that I rather use an arrow function, but does not allow ES6 features.

---

*[qutebrowser/javascript/stylesheet.js, line 105 at r1](https://reviewable.io:443/reviews/qutebrowser/qutebrowser/3212#-KxgGWIbkz_CflTlPlMV:-Kxi9rp6nSY6L5SJx84M:bjvqifc) ([raw file](https://github.com/qutebrowser/qutebrowser/blob/2fe1a1db89f97c23b3ca1aeac8fb33f4d8c5f940/qutebrowser/javascript/stylesheet.js#L105)):*
<details><summary><i>Previously, sandrosc (Sandro) wrote…</i></summary><blockquote>

For which elements should `check_style` be called? `createNodeIterator` allows for some filtering, but could we also define a list of elements to be checked?
</blockquote></details>

It should be called for all processing instructions (to check for an `xml-stylesheet` instruction) and elements (to check for an XHTML or SVG element, which both have associated default styles). The `NodeIterator` is used to check existing nodes, and the `MutationObserver` to check any nodes added after this function returns. Filtering does give a small performance improvement, so I will add it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/qutebrowser/qutebrowser/3212)*

**Comment by The-Compiler** [comment]
@ulrikdem I probably should've mentioned this earlier (sorry for not doing so!), but the eslint rules/config are absolutely open to discussion!

The config is at `qutebrowser/javascript/.eslintrc.yaml` - the reason it's quite picky is that I'd like to keep the `extends: "eslint:all"` which turns on all checkers, and then selectively turn them off when they're in the way.

Feel free to e.g. turn off `no-bitwise` globally there (I don't think it makes sense to have this check), and if anything else comes up which conflicts with common sense, please let me know!

I also *think* we should be able to migrate to ECMAscript 6 now that we dropped legacy QtWebKit support, but that's probably material for a separate PR.

---

Reviewed 1 of 2 files at r2.
Review status: 3 of 4 files reviewed at latest revision, 6 unresolved discussions.

---

*[qutebrowser/browser/webengine/webenginesettings.py, line 167 at r1](https://reviewable.io:443/reviews/qutebrowser/qutebrowser/3212#-KxiAL-Gecit57wx8Pj6:-KxidyfhoFVaqTgAd74_:b-63sp3m) ([raw file](https://github.com/qutebrowser/qutebrowser/blob/2fe1a1db89f97c23b3ca1aeac8fb33f4d8c5f940/qutebrowser/browser/webengine/webenginesettings.py#L167)):*
<details><summary><i>Previously, ulrikdem (Ulrik de Muelenaere) wrote…</i></summary><blockquote>

It is too late. However, wouldn't the JS file be read once per tab in `WebEngineTab._init_js` but only when the configuration is changed where it is currently? And injected once into each page in either case?
</blockquote></details>

Hmm, you're right. I suppose the initialization could be moved from there to `qutebrowser/utils/javascript.py` or so, to work on the profile instead. I guess it's fine to keep things as-is in this PR then.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/qutebrowser/qutebrowser/3212#-:-KxieD7mUTVqlpyh9NO5:bpdnshl)*

**Comment by The-Compiler** [comment]
Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[qutebrowser/javascript/stylesheet.js, line 27 at r3](https://reviewable.io:443/reviews/qutebrowser/qutebrowser/3212#-Ky0F5iO_EkToupdB__t:-Ky0F5iO_EkToupdB__u:b-6wxy2w) ([raw file](https://github.com/qutebrowser/qutebrowser/blob/95b41b311f6199da976b7bafbd3148d2aa35287f/qutebrowser/javascript/stylesheet.js#L27)):*
> ```JavaScript
>     }
> 
>     var funcs = {};
> ```

Quick update: With #3217, qutebrowser now uses ES6 and disallows `var` - so you should use [let](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/let)/[const](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/let) instead of `var` everywhere, and can use [most other](https://github.com/qutebrowser/qutebrowser/issues/3216#issue-269724431) ES6 features if you want.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/qutebrowser/qutebrowser/3212)*

**Comment by The-Compiler** [comment]
:lgtm: now! The only thing missing now are tests for all the funny edge cases, which @jgkamat wanted to take care of when he finds the time. 

---

Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/qutebrowser/qutebrowser/3212#-:-KyJun98HVum2H6-1Z5C:b-gb5x81)*

**Comment by jgkamat** [comment]
Hey, I think I found a bug in this while testing it. When loading a xml file [such as this site](https://www.w3schools.com/xml/note.xml), I get this error:

```
Traceback (most recent call last):
  File "/home/jay/Code/qutebrowser/tests/unit/javascript/conftest.py", line 100, in javaScriptConsoleMessage
    pytest.fail("[{}] js console ({}:{}): {}".format(level, source, line, msg))
  File "/home/jay/Code/qutebrowser/.tox/py35/lib/python3.5/site-packages/_pytest/outcomes.py", line 85, in fail
    raise Failed(msg=msg, pytrace=pytrace)
Failed: [2] js console (userscript:_qute_stylesheet:101): Uncaught TypeError: mutation.addedNodes[Symbol.iterator] is not a function
```

Can you reproduce? I guess this testing this actually works :smile:

**Comment by ulrikdem** [comment]
I can't reproduce this. Could you try with the following patch?

```diff
diff --git a/qutebrowser/javascript/stylesheet.js b/qutebrowser/javascript/stylesheet.js
index 13741cf41..b1cdeb26e 100644
--- a/qutebrowser/javascript/stylesheet.js
+++ b/qutebrowser/javascript/stylesheet.js
@@ -96,8 +96,9 @@ window._qutebrowser.stylesheet = (function() {
         }
         const style_observer = new MutationObserver((mutations) => {
             for (const mutation of mutations) {
-                for (const added of mutation.addedNodes) {
-                    if (check_style(added)) {
+                const nodes = mutation.addedNodes;
+                for (let i = 0; i < nodes.length; ++i) {
+                    if (check_style(nodes[i])) {
                         create_style();
                         style_observer.disconnect();
                         return;
```

**Comment by jgkamat** [comment]
This solves the error for me, but now xml documents are having style applied to them (is this expected, or not, I'm not sure). FWIW, I use qtwebengine 5.7.1, so maybe that type of loop is a newer addition?

2017-11-09-114106_1920x1080_scrot

**Comment by ulrikdem** [comment]
It must be the QtWebEngine version. I have pushed the fix.

My intention is that the user stylesheet is applied after the document is rewritten to show the XML tree, as in your screenshot.

**Comment by The-Compiler** [comment]
Running the [acid3 tests](http://acid3.acidtests.org/) with this change, it looks like there's a new failure introduced by this (click above the logo to get the test information):

Before:

```
$ python3 -m qutebrowser --temp-basedir  --debug --logfilter js http://acid3.acidtests.org/
07:55:04 DEBUG    js         shared:javascript_log_message:136 [userscript:_qute_stylesheet:6] Uncaught TypeError: Cannot read property 'appendChild' of undefined
07:55:04 DEBUG    js         shared:javascript_log_message:136 [userscript:_qute_stylesheet:6] Uncaught TypeError: Cannot read property 'appendChild' of undefined
07:55:05 DEBUG    js         shared:javascript_log_message:136 [userscript:_qute_stylesheet:6] Uncaught TypeError: Cannot read property 'appendChild' of undefined
07:55:08 DEBUG    js         shared:javascript_alert:110 alert: Failed 1 tests.
Test 26 passed, but took 45ms (less than 30fps)
Test 69 passed, but took 37 attempts (less than perfect).
Test 71 failed: expected '1' but got '2' - wrong number of children in HEAD (first test)
Total elapsed time: 1.58s
```

with this change (but no custom stylesheet set):

```
$ python3 -m qutebrowser --temp-basedir  --debug --logfilter js http://acid3.acidtests.org/
07:55:31 DEBUG    js         shared:javascript_alert:110 alert: Failed 2 tests.
Test 26 passed, but took 44ms (less than 30fps)
Test 65 passed, but took 51ms (less than 30fps)
Test 69 passed, but took 34 attempts (less than perfect).
Test 71 failed: expected '2' but got '3' - wrong number of children in HTML (first test)
Test 72 failed: expected '20' but got '10' - change failed to take effect
```

New failure: `Test 72 failed: expected '20' but got '10' - change failed to take effect`

So far, I've been unable to find out what those tests actually are though...

**Comment by The-Compiler** [comment]
I merged this for now, let's take care of the Acid failure whenever we take care of #3295 :wink:

Thanks a lot - this is the kind of JS code I definitely couldn't have written myself, and it makes things so much better! 👍
